pkg/stanza/fileconsumer/internal/reader/reader.go

```diff
@@ -209,15 +212,47 @@ func (r *Reader) readContents(ctx context.Context) {
 			r.FileAttributes[attrs.LogFileRecordNumber] = r.RecordNum
 		}
 
-		err = r.emitFunc(ctx, emit.NewToken(token, r.FileAttributes))
-		if err != nil {
-			r.set.Logger.Error("failed to process token", zap.Error(err))
+		tokens = append(tokens, emit.NewToken(copyBody(token), copyAttributes(r.FileAttributes)))
+
+		if r.maxBatchSize > 0 && len(tokens) >= r.maxBatchSize {
+			err := r.emitFunc(ctx, tokens)
+			if err != nil {
+				r.set.Logger.Error("failed to emit token", zap.Error(err))
+			}
+			tokens = tokens[:0]
+			r.Offset = s.Pos()
 		}
+	}
 
+	if len(tokens) > 0 {
+		err := r.emitFunc(ctx, tokens)
+		if err != nil {
+			r.set.Logger.Error("failed to emit token", zap.Error(err))
+		}
```

Comment:
If I understand correctly, this if statement only applies to the situation where we've hit EOF, so WDYT about moving it to where we detect that situation, rather than breaking the loop and falling into this.

Reply:
Oh, I think you're right. I think I assumed that this can also happen when the [context is canceled](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36663/files#diff-0788680cd9b217cf2e6aa57e88309c12f83aa22c0680c25585c22d620ca0fa35R188) (e.g. when collector receives SIGINT). But when the context is canceled, the method `return`s instead of `break`ing out of the loop, which I suppose is an OK thing to do. I'll take another look at it.

Reply:
Moved the code in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36663/commits/961789e97fcd6a202c1b43a96cbec3fcc8ea6fce.